Improve JavaScript filetype plugin

romainl opened the pull request:
- set missing &suffixesadd
- set missing &suffixes
- set missing &define
- set missing &path to a minimal default value
- load appropriate compiler for &makeprg and &errorformat
- modify $PATH
- add missing b:match_words
- add patterns to b:browsefilter

romainl committed:
Improve JavaScript filetype plugin

- set missing &suffixesadd
- set missing &suffixes
- set missing &define
- set missing &path to a minimal default value
- load appropriate compiler for &makeprg and &errorformat
- modify $PATH
- add missing b:match_words
- add patterns to b:browsefilter

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
 if has("gui_win32")
-    let  b:browsefilter="Javascript Files (*.js)\t*.js\n" .
-		\	"All Files (*.*)\t*.*\n"
+    let  b:browsefilter="Javascript Files (*.js, *.jsx, *.es, *.es6, *.cjs, *.mjs, *.jsm, *.vue, *.json)\t*.js;*.jsx;*.es;*.es6;*.cjs;*.mjs;*.jsm;*.vue;*.json\n" .
```
This would probably be better divided into groups for "JavaScript Files", "JavaScript Modules", "JSON Files", "JSX Files", and "Vue Template Files".

romainl replied:
Good point. Will do.

romainl replied:
Done.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+    let $PATH = s:bin_dir .. ':' .. $PATH
+endif
+unlet s:bin_dir
+
```
I think this is more likely to annoy than it is to help and should really be set up externally.

No other ftplugin is modifying PATH.

romainl replied:
Without this it is very hard to use linters or formatters as it is customary to include them as project dependencies.

Maybe it would be possible to perform path resolution at the formatter/compiler level?

dkearns replied:
This is the purpose of the system PATH environment variable.  A quick search suggests that this would be the only place in the entire runtime distribution that $PATH was modified.

romainl replied:
I know what's the purpose of `$PATH` and adding specific directories to `$PATH` fits that purpose entirely. Again, without this locally-installed linters and formatters can't be called.

Please suggest a better approach that, ideally, doesn't require writing a hundred lines of documentation that the user will never read.

> A quick search suggests that this would be the only place in the entire runtime distribution that $PATH was modified.

JavaScript is one of the rare ecosystem where that might be useful.

dkearns replied:
By "system PATH environment variable" I meant it should be properly configured in the environment from which Vim is started.  This is the better approach.

It's unacceptable to be adding **relative paths** to **the head** of Vim's $PATH **without the user's express consent** and being the only ftplugin to do so makes it extra surprising and dangerous.  

Even in an ideal world, the PATH component you're prepending, being relative, is really only useful when Vim's working directory is fixed.

I suppose if you want to go all in on Node.js you could have the compilers use npx but that wouldn't really be my preference.  It's probably a rather slippery slope.

> JavaScript is one of the rare ecosystem where that might be useful.

JavaScript is not special. ;)

romainl replied:
> By "system PATH environment variable" I meant it should be properly configured in the environment from which Vim is started. This is the better approach.

A better approach that prevents use of linters and formatters that are installed locally.

> Even in an ideal world, the PATH component you're prepending, being relative, is really only useful when Vim's working directory is fixed.

It is absolute.

> I suppose if you want to go all in on Node.js you could have the compilers use npx but that wouldn't really be my preference. It's probably a rather slippery slope.

I don't want to go that way, hence this.

> JavaScript is not special. ;)

The language isn't. The ecosystem and the established practices are.

romainl replied:
@dkearns, ESLint and Standard are currently the most widely used linters in JavaScript land and beyond (ESLint is now the official TypeScript linter). You are right that people often use plugins to interact with those external tools but that has more to do with the state of abandonment of the JavaScript runtime than with technical limitations.

Now that I have added proper "compilers" for ESLint and Standard to `$VIMRUNTIME/compiler/` all that is left is to make them usable in the most seamless way possible.

But we need a quick reality check, first: those tools can certainly be installed globally but they are more often than not installed *locally*, as part of project dependencies. This means that:

- if the tool is installed globally it is likely to be somewhere in `$PATH`,
- if the tool is installed locally it is likely to not be in `$PATH`.

The second case is extremely easy to handle with minimal overhead and that's what I did here:

1. find the absolute path to the parent `node_modules/.bin`,
2. check if it is already in `$PATH`,
3. if not, prepend it to `$PATH`.

The idea is to make sure locally-installed tools are as accessible and usable as if they were installed globally. It is totally transparent and requires no user action.

With that setup and the `:compiler` commands in the ftplugin, here is what the user would have to do before being able to use of the built-in `eslint` compiler:

1. edit a JavaScript file.

By contrast, here is what the user would have to do before being able to use of the built-in `eslint` compiler:

1. do `$ PATH=<absolute path to node_modules/.bin/>:$PATH` either before starting Vim or from Vim itself,
2. do `:compiler eslint` in every relevant buffer, or do `:compiler! eslint` for *every* buffer (with caveat).

Yes, that's not a really a common strategy in `$VIMRUNTIME/` but that strategy is, in my opinion, a significant upgrade to the existing status quo.

dkearns replied:
Sorry, I lazily misremembered some of the details.  I'm reviewing this without access to a dev box and I should have reread the diff before commenting again.

> It is absolute.

It is but it's a dynamically generated absolute path based on the current file's location and a directory presence test.  This seems very unusual (possibly unique) at best and dangerous at worst and I'm not convinced it is a reasonable way to solve the perceived problem.

As it stands, editing a JS file anywhere on the filesystem potentially modifies `PATH` and does so by favouring commands installed in any discovered `node_modules/.bin` directory.  

Users have to be able to know what is going to be executed when running a shell command from within Vim yet this now occurs within a possibly surreptitiously modified environment.  It's a security concern much like adding `.` to the head of `PATH` but arguably worse with the further path indirection and the fact it wasn't set by the user themselves.

It would be a little better to perform the directory search relative to the cwd.  At least the user has some commitment to operating in the project hierarchy having started Vim from there but I suspect that doesn't match all your use cases.

Anyway, if we were to include something like this I think it would have to be guarded by an opt-in user config variable.

Some implementation observations while I'm reading the diff:

- It seems that the cwd will be prefixed to `PATH` (if it's not already there) whenever `node_modules/.bin` is not found because IIRC `fnamemodify('', ':p')` returns the current directory.
- The pattern test should be tightened up to match only against the full length of a path component allowing for optional `/`s and to escape the `.` metacharacter.  It would be more elegant to split the path and search the resultant list but I like a regex test as much as anyone.
- It might be better to use empty() to test the path rather than len()
- The value for `PATH` needs to be reset using the undo_ftplugin mechanism.

However, wouldn't it be better to just make use of a call to `npm` itself to determine the bin directory?  One would hope it was a robust solution and not prohibitively slow.

> But we need a quick reality check, first: those tools can certainly be installed globally but they are more often than not installed _locally_, as part of project dependencies. This means that:
> 
>     * if the tool is installed globally it is likely to be somewhere in `$PATH`,
> 
>     * if the tool is installed locally it is likely to not be in `$PATH`.

I find it difficult to imagine working in a development environment (Node.js related or otherwise) where `PATH` isn't set appropriately to pick up these locally installed tools if I intended using them.  What environment is the user working in where setting `PATH` for a project is such an onerous task?

It's, generally, the assumption for all external tools utilised from within Vim.  Some compilers also have similar expectations for other environment variables like `CLASSPATH` while others still don't even bother trying to set `makeprg` and leave that up to the user.

> The idea is to make sure locally-installed tools are as accessible and usable as if they were installed globally. It is totally transparent and requires no user action.

I understand what you're trying to achieve but not the motivation.  Who exactly is this feature targeting?  New developers, new Vim users, people who want Vim to behave like an IDE, or some combination of these?  It's not clear to me but I'm open to the notion that I have a huge blind spot.

Why do you think this is a project configuration issue that should be resolved in Vim and presumably every other program that needs to run these tools and if it's more than a project configuration issue then surely it's a Node.js problem that should be solved there?  A quick web search reveals a plethora of suggested solutions.

I agree that they have never handled this very gracefully and certainly worse than other language environments but on further investigation it seems that `npx` is, indeed, the blessed solution at the moment.  The docs for both `eslint` and `standard`, for example, explicitly specify its use in this situation.  So I think, contradicting what I said earlier, that having the compilers themselves utilise `npx` _is_ probably the best solution.

> By contrast, here is what the user would have to do before being able to use of the built-in `eslint` compiler:
>     1. do $ PATH=<absolute path to node_modules/.bin/>:$PATH either before starting Vim or from Vim itself,

This, or something similar, has to be done once per project at creation time.

>     2. do `:compiler eslint` in every relevant buffer, or do `:compiler! eslint` for _every_ buffer (with caveat).

No, this would generally be automated using existing Vim mechanisms; there's lots of ways to do it.

> Yes, that's not a really a common strategy in `$VIMRUNTIME/` but that strategy is, in my opinion, a significant upgrade to the existing status quo.

The bone of contention...

I think there's more value in consistency of behaviour across all ftplugins even if that might seem overly conservative in some instances.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+setlocal path-=node_modules
+setlocal path-=/usr/include
+setlocal path-=**
+
```
Why are `node_modules` and `**` being removed?  Where were they added, and if they were, why remove them?

romainl replied:
> Where were they added?

Somewhere in the user's `vimrc`.

> Why remove them?

Because they make everything slower. `**` is a very bad idea that perverts the very idea of `&path`, `node_modules` is black hole, and `/usr/include` is irrelevant.

dkearns replied:
It seems rather heavy-handed to be removing path components when we don't even offer an improved language specific default.

I wouldn't argue with removing `/usr/include` but `node_modules` seems like an entirely reasonable component and if they intentionally specified something like `**` then I can't seen any JavaScript related reason to remove it.

romainl replied:
- `**` is too greedy and completely defeats the purpose of `&path` which is to explicitly list interesting directories from which to look for files. `dist/` or `coverage/` are totally irrelevant so they shouldn't be searched.
- And the effect of `**` is aggravated by `node_modules/` which is, again, a black hole for Vim's search algorithm. That's as "JavaScript" as it gets, IMO.

I could settle for only removing `/usr/include` and letting the user shoot themselves in the foot if they so choose.

dkearns replied:
This seems like highly personal opinion.  If the user wants these in their path leave them there.  They almost certainly know better than us.

There are an infinite number of potentially undesirable path components.  Removing `**` is totally arbitrary.  You may as well be arguing for its removal from Vim itself by saying it's incompatible with `path`.

E.g., why aren't you removing `node_modules/**`?

Irrespective, it makes much more sense assuming these are real problems, to set a complete path value rather than remove components.

To be clear the major problem I have with this is that it's almost certainly just line noise.  The chance of someone accidentally including these components and being grateful they were removed is very small.

romainl replied:
Unfortunately, users put `**` in their `&path` because they were told to do it by other misinformed users. `set path+=**` is widely used despite it breaking lots of commands. This is only an attempt to alleviate all too common problems.

Anyway, as I said I will just remove it.

romainl replied:
DONE in commit 850c35f

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
 
-" Change the :browse e filter to primarily show Java-related files.
+" Change the :browse e filter to primarily show JavaScript-related files.
 if has("gui_win32")
```
```suggestion
if (has("gui_win32") || has("gui_gtk")) && !exists("b:browsefilter")
```

romainl replied:
Good point.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+setlocal path-=/usr/include
+setlocal path-=**
+
+" Matchit configuration
```
I don't think Matchit is a very good match for JavaScript syntax.

The only true pair here is `do/while`.  `function` can have no, or multiple, `return` statements and `default` and `finally` are optional in their pair groups.  These badly break the contract for `%`.

You'll notice that similar languages like C, Java, C# etc don't make use of it.

romainl replied:
You are probably right.

dkearns replied:
Leaving do/while as an easter egg kind of appeals to me.

romainl replied:
Sounds fine by me. Will do.

dkearns replied:
As an aside, have you come across a plugin that handles this sort of navigation for C-like syntax?

romainl replied:
> As an aside, have you come across a plugin that handles this sort of navigation for C-like syntax?

I have no idea what you mean.

romainl replied:
DONE in commit 850c35f (and further fixed in 2b21d305d).

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+            \ .. '\<switch\>:\<case\>:\<default\>,'
+            \ .. '\<if\>:\<else\>,'
+            \ .. '\<try\>:\<catch\>:\<finally\>,'
+            \ .. '<\(\w\+\):</\1>'
```
JSX has its own filetype (javascriptreact.vim) so this final pattern should be moved there.

The start pattern is missing the closing `>`

romainl replied:
- React is not the only library/framework that makes use of `XML`-like syntax.
- Omitting the closing angle bracket makes the pattern simpler.

dkearns replied:
>     * React is not the only library/framework that makes use of `XML`-like syntax.
I thought they could be assigned the `javascriptreact` filetype? 
>     * Omitting the closing angle bracket makes the pattern simpler.
It's broken when nested elements have a name prefix matching the name of a parent element. E.g.,
`<s>...<span>...</span>...</s>`

romainl replied:
> It's broken when nested elements have a name prefix matching the name of a parent element. E.g.,
>     <s>...<span>...</span>...</s>

Not in my real-life experience. In the sample below, `%` correctly jumps between `<StyledBanners>` and `</StyledBanners>` or between `<Space>` and `</Space>` :

    <StyledBanners>
      <Space>
        <Station>
          <Sacrifice>
            {foo}
          </Sacrifice>
        </Station>
      </Space>
    </StyledBanners>

The setting is indeed broken with your sample but not in the way you describe:

`<s>...</s>` doesn't work but `<span>...</span>` works. Adding any legal character to `<s>` and `</s>` makes it work as expected, though. This is unexpected.

dkearns replied:
In my example, the "s" prefix (if you will) of the span element matches the name, "s", of the parent strikethrough element.  This is not the case in your example but would be so if the Space element was named StyledBannersSpace.

Adding any character other than a "p" to the parent element name makes it work. This is expected.

Regardless, the original pattern is clearly underspecified.  Is there a reason not to just reuse the relevant parts of the XML ftplugin's value for b:match_words?

romainl replied:
> Regardless, the original pattern is clearly underspecified. Is there a reason not to just reuse the relevant parts of the XML ftplugin's value for b:match_words?

I could do that.

andymass replied:
The `\<if\>:\<else\>,'` stuff is not really conforming.  matchit's b:match_words syntax is `start:middle:end` or `start:end`.  But `else` doesn't end a block so it's not likely to work correctly

romainl replied:
@andymass, groups can have two or more patterns so `\<if\>:\<else\>` is a perfectly valid group that works as expected on:

    if () {
    } else {
    }

but it can't handle more than one `else`, which is why it will go.

romainl replied:
DONE in commit 850c35f (and further fixed in 2b21d305d)

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+endif
+
+" Set 'define' to a comprehensive value
+let &define = '\('
```
This should use `&l:define` so the value is buffer local.

romainl replied:
Agreed.

romainl replied:
Done.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+" Use standard for :make if applicable
+if executable('standard')
+    compiler standard
+endif
```
In general, I think, compiler settings are excluded from ftplugins unless there is a very clear default choice.  There's currently only three ftplugins that set a compiler.

If they were included they'd need to be wrapped in tests for the `current_compiler` variable so that they don't overwrite a previously set compiler.

romainl replied:
- Including them here means less fiddling and a better out-of-the-box experience.
- Testing for `current_compiler` is already handled at the compiler level.

dkearns replied:
>     * Including them here means less fiddling and a better out-of-the-box experience.
I don't feel strongly about it but that's the way it's done and I don't see a lot of value in departing from the project norms.  However, if you were keen to offer a default it might be worth adding a `javascript_default_compiler` config variable like the Ada ftplugin does.
>     * Testing for `current_compiler` is already handled at the compiler level.
It's being tested in a different context there.  That's so a user compiler can block the default compiler from being loaded.  The `:compiler` command actually deletes it when called from the ftplugin.

romainl replied:
> I don't feel strongly about it but that's the way it's done and I don't see a lot of value in departing from the project norms. However, if you were keen to offer a default it might be worth adding a javascript_default_compiler config variable like the Ada ftplugin does.

The problem, here, is that JavaScript is not a compiled language that comes with a default or recommended compiler. Defaulting to GNAT makes sense in Ada but defaulting to anything doesn't make sense in JS-land.

The right "compiler" is always project-specific and generally installed locally so having a default compiler, user-configurable or not, doesn't make much sense to me.

> It's being tested in a different context there. That's so a user compiler can block the default compiler from being loaded. The `:compiler` command actually deletes it when called from the ftplugin.

That seems strange but I'll take your word for this.

dkearns replied:
It doesn't really make sense, I know.  I was just suggesting a config variable as you don't seem to like the standard way(s) of selecting a compiler plugin.  You are, however, currently favouring `eslint` over `standard` in this setup. 

I have at least a half dozen other compilers that I've been meaning to clean up and send upstream...for some years.  I suspect most people are actually using one of the linting plugins that provide these compilers by default.

As for testing `current_compiler` you can check the C source for the details if you're interested or just go with the boiler plate.  Don't fight the cargo cult.

romainl replied:
> You are, however, currently favouring `eslint` over `standard` in this setup.

No. If you have both, `standard` is set last. If you only have `eslint`, `eslint` is set. And if you only have `standard`, `standard` is set. ESLint being a dependency of Standard, that behaviour makes perfect sense.

> I have at least a half dozen other compilers that I've been meaning to clean up and send upstream...for some years.

I already have pushed `eslint.vim` and `standard.vim`. What others do you have in mind? JSHint? Maybe Babel?

`:compiler` is not a special command. Like all its friends it can be used in mappings, other commands, scripts, autocommands, etc. Or in a ftplugin.

dkearns replied:
> > You are, however, currently favouring `eslint` over `standard` in this setup.
> 
> No. If you have both, `standard` is set last. If you only have `eslint`, `eslint` is set. And if you only have `standard`, `standard` is set. ESLint being a dependency of Standard, that behaviour makes perfect sense.

Right, the other way around. :)

Irrespective, it's not that easy to get the defaults right even with only two compilers.  Currently, for example, if you have `standard` globally installed and `eslint` locally installed it'll set the compiler to `standard` which doesn't seem like the desired result.

None of this particularly matters because the desired compiler can simply be setup as usual but unless it's going to be the correct default for most users I don't see the point.  There's also the potential downside of building an expectation among naive users that the default will automagically be correct and having to field support requests when, for whatever reason, it no longer is.

> > I have at least a half dozen other compilers that I've been meaning to clean up and send upstream...for some years.
> 
> I already have pushed `eslint.vim` and `standard.vim`. What others do you have in mind? JSHint? Maybe Babel?

Yes and Flow and some others like Rhino, SpiderMonkey etc.

> `:compiler` is not a special command. Like all its friends it can be used in mappings, other commands, scripts, autocommands, etc. Or in a ftplugin.

I was referring to the need to wrap it in a test for `current_compiler` in the ftplugin.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -1,19 +1,20 @@
+" Language:     Javascript
+" Maintainer:   Doug Kearns <dougkearns@gmail.com>
+" Last Change:  2020 Jun 23
+" URL:          http://gus.gscit.monash.edu.au/~djkea2/vim/ftplugin/javascript.vim
```
```suggestion
```

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +27,63 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+            \ .. '\|^\s*\(\ze\i\{-}\)\(([^)]*).*{$\|\s*[:=,]\)'
+            \ .. '\|^\s*\(export\s\{-}\|export\s\{-}default\s\{-}\)*\(var\|let\|const\|function\)'
+            \ .. '\|\(\<as\>\)'
+            \ .. '\)'
```
I'll test this when I get a bit more time but just looking at it for now:

-  the entire pattern doesn't need to be grouped with parens nor does `\<as\>`
- whenever you've used `\{-}` you meant `\+`
- it's not matching class definitions or generator functions

romainl replied:
- I will look into the parentheses issues.
- No, I meant `\{-1,}`. Will fix this.
- Generator functions are matched but classes are not, I'll make sure it does.

dkearns replied:
>     * I will look into the parentheses issues.
> 
>     * No, I meant `\{-1,}`. Will fix this.
That always matches the same thing as `\+` in the contexts it's currently used and reads a bit like a regex double negative to me.
>     * Generator functions are matched but classes are not, I'll make sure it does.
Sorry I forgot it skips non-identifier chars rather than just whitespace.

If you're working on this now I'll wait for the next update to test it.

romainl replied:
@dkearns, commit 8f245c29a:

- adds class definitions,
- replaces all `\{-}` with `\+`
- removes some unnecessary parentheses.

FWIW, the value supports the following in its current state:

    function foo
    (function foo
    async function foo
    (async function foo
    function * foo
    var foo
    let foo
    const foo
    static foo
    async foo
    get foo
    set foo
    this.foo
    identifier.foo
    foo()
    foo =
    foo,
    export let foo
    export var foo
    export const foo
    export function foo
    export class foo
    export default let foo
    export default const foo
    export default var foo
    export default function foo
    export default class foo
    import ... as foo
    export ... as foo
    * foo

NOTE: Vim is incapable of finding multiple definitions on the same line.

dkearns reviewed (commented):
Thanks.  See inline comments.

Now that you've added `define` we should probably look at providing a reasonable value for the `include` option too.

romainl commented:
> Thanks. See inline comments.
> 
> Now that you've added `define` we should probably look at providing a reasonable value for the `include` option too.

I have a comprehensive `include` but the nature of JavaScript and its ecosystem, as well as Vim limitations, make proper include search quite a challenge. A working `include` is not enough, we also need a working `includeexpr`—which I also in the works—that works efficiently even on large trees but that's currently impossible (directory handling is an issue, the way search is performed is another one).

I don't think we can add an `include` right now.

romainl committed:
Update runtime/ftplugin/javascript.vim

Co-authored-by: Doug Kearns <dougkearns@gmail.com>

romainl committed:
Update runtime/ftplugin/javascript.vim

Co-authored-by: Doug Kearns <dougkearns@gmail.com>

romainl committed:
b:browsefilter is better organised

romainl committed:
Make 'define' local

romainl commented:
> Thanks. See inline comments.
> 
> Now that you've added `define` we should probably look at providing a reasonable value for the `include` option too.

I actually have had a perfectly working `&include` for a while:

    ^\s*[^/]\+\(from\|require(\)\s*['"]\ze

but module resolution is complicated in JavaScript and Vim:

- relative imports are relative to the current file,
- `./foo` could match `./foo/index.js`, `./foo.js`, or even `./foo/foo.js` via `./foo/index.js`,
- given the path `./foo`, Vim considers the directory `./foo/` to be a valid target so it stops there,
- how to deal with aliases?
- etc.

Because of that, we also *need* a working `&includeexpr` that handles every corner case. I have a reasonably smart one that deals with local imports, aliased imports, and `node_modules/` imports quite gracefully, but now that we have a working `&include` and a working `&includeexpr`, we have to face another problem: modern JavaScript projects tend to have large "main" files like `App.js` with dozens of imports, themselves with dozens of imports and so on and, speaking from experience, doing `[<C-d>` in such a file can easily take tens of seconds to reach its target because of the sheer size of the tree (essentially the whole application). And of course, the smaller the import tree, the better the performance so you have an include search that works very well in smaller components while being essentially useless in larger core files.

And all that heavily depends on a properly set `&path`, which is another can of worms.

And then you have `gf` and friends who don't even attempt to use `&includeexpr` if they find a matching directory. Those need to be overridden, too.

As far as I'm concerned, include search in JS is not a solved problem and it is way too early to introduce a working but ultimately deceiving `&include`.

romainl committed:
Handle PR feedbacks

- only remove /usr/include from &path
- only handle do...while and XML/HTML/*SX tags in b:match_words

romainl committed:
Fixed hasty b:match_words

romainl committed:
Refined &define value

SUPPORTED:
function foo
(function foo
async function foo
(async function foo
function * foo
class foo
var foo
let foo
const foo
static foo
async foo
get foo
set foo
this.foo
identifier.foo
foo()
foo =
foo,
export let foo
export var foo
export const foo
export function foo
export class foo
export default let foo
export default const foo
export default var foo
export default function foo
export default class foo
import ... as foo
export ... as foo
* foo

NOT SUPPORTED:
multiple definitions per line (Vim limitation)

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +26,65 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+                \ "JavaScript Modules (*.es, *.es6, *.cjs, *.mjs, *.jsm)\t*.es;*.es6;*.cjs;*.mjs;*.jsm\n" .
+                \ "Vue Templates (*.vue)\t*.vue\n" .
+                \ "JSON Files (*.json)\t*.json\n" .
+                \ "All Files (*.*)\t*.*\n"
```
Minor style point - these lines should be updated to use the new `..` concatenation operator as done elsewhere.

romainl replied:
Agreed, I did it elsewhere, and will fix these ASAP.

romainl replied:
Done.

dkearns reviewed (commented):
> I don't think we can add an `include` right now.

In principle I don't have any objection to including a partially working feature if it's clear it has potential to be improved and it does no harm.  That line, however, isn't always clear and as you say incorrect results can be rather frustrating.

I'll leave that up to you.

romainl commented:
> As it stands, editing a JS file anywhere on the filesystem potentially modifies PATH and does so by favouring commands installed in any discovered node_modules/.bin directory.

Only if that directory is found in a parent directory of the current file. This is, for better or for worse, how it is supposed to work in JS land.

> Users have to be able to know what is going to be executed when running a shell command from within Vim yet this now occurs within a possibly surreptitiously modified environment. It's a security concern much like adding . to the head of PATH but arguably worse with the further path indirection and the fact it wasn't set by the user themselves.

> It would be a little better to perform the directory search relative to the cwd. At least the user has some commitment to operating in the project hierarchy having started Vim from there but I suspect that doesn't match all your use cases.

- Many Windows users work with the current directory set to `C:`.
- Opening a file from a different project in a new tab without doing `:lcd` or `:tcd` first would have bizarre consequences.
- Working in a monorepo would imply either starting separate sessions in the different sub-projects or, again, juggling with `:lcd`/`:tcd`.

> Anyway, if we were to include something like this I think it would have to be guarded by an opt-in user config variable.

I would prefer not to but that would be acceptable.

> It seems that the cwd will be prefixed to PATH (if it's not already there) whenever node_modules/.bin is not found because IIRC fnamemodify('', ':p') returns the current directory.

Agreed. Done.

> The pattern test should be tightened up to match only against the full length of a path component allowing for optional /s and to escape the . metacharacter. It would be more elegant to split the path and search the resultant list but I like a regex test as much as anyone.

I will have to think about it a little more.

> It might be better to use empty() to test the path rather than len()

Agreed. Done.

> The value for PATH needs to be reset using the undo_ftplugin mechanism.

Agreed, and it is.

However, wouldn't it be better to just make use of a call to npm itself to determine the bin directory? One would hope it was a robust solution and not prohibitively slow.

`npm` and `node` are horrendously slow but we can do it async.

> I find it difficult to imagine working in a development environment (Node.js related or otherwise) where PATH isn't set appropriately to pick up these locally installed tools if I intended using them. What environment is the user working in where setting PATH for a project is such an onerous task?

An environment where they don't have to do that because their editor/IDE or their plugins are smart enough to do it for them.

Adding dozens of non-globally-significant paths to the globally significant `$PATH` makes no sense whatsoever. Having your tools handle `$PATH` dynamically for you within the context of editing makes a lot more sense.

> It's, generally, the assumption for all external tools utilised from within Vim. Some compilers also have similar expectations for other environment variables like CLASSPATH while others still don't even bother trying to set makeprg and leave that up to the user.

> I understand what you're trying to achieve but not the motivation. Who exactly is this feature targeting? New developers, new Vim users, people who want Vim to behave like an IDE, or some combination of these? It's not clear to me but I'm open to the notion that I have a huge blind spot.

People who want to benefit from Vim's built-in features with as few hassles as possible. `:make %` is theoretically easy but too hard in practice because of the setup involved. Same for `gq`. Same for `:find`. Same for `[<C-d>`. And the list goes on.

The cost is so hard that people turn too easily to third-party alternatives… and never learn how to use the built-in features because no time was ever invested into making the default ftplugin-driven experience better for JavaScript. I want to change that.

> Why do you think this is a project configuration issue that should be resolved in Vim and presumably every other program that needs to run these tools and if it's more than a project configuration issue then surely it's a Node.js problem that should be solved there? A quick web search reveals a plethora of suggested solutions.

Because it so much easier to do at the editor level.

> I agree that they have never handled this very gracefully and certainly worse than other language environments but on further investigation it seems that npx is, indeed, the blessed solution at the moment. The docs for both eslint and standard, for example, explicitly specify its use in this situation. So I think, contradicting what I said earlier, that having the compilers themselves utilise npx is probably the best solution.

`npx` is a cool tool for one-off tasks but the resolution adds ~2s to each command so it is unusable for lint-on-write scenarios, for example, or even the occasional `gq`.

> 1. do $ PATH=<absolute path to node_modules/.bin/>:$PATH either before starting Vim or from Vim itself,

> This, or something similar, has to be done once per project at creation time.

And stored where? I have 149 JS projects on this machine, there is no way I would add all of them to my $PATH when **NONE** of those 149 `node_modules/.bin` would be relevant outside of their specific project. That idea makes no sense at all.

> 2. do `:compiler eslint` in every relevant buffer, or do `:compiler! eslint` for _every_ buffer (with caveat).

> No, this would generally be automated using existing Vim mechanisms; there's lots of ways to do it.

Like… the proposed mechanism.

> Yes, that's not a really a common strategy in $VIMRUNTIME/ but that strategy is, in my opinion, a significant upgrade to the existing status quo.

> The bone of contention...

> I think there's more value in consistency of behaviour across all ftplugins even if that might seem overly conservative in some instances.

I think there is more value in providing as much value as possible to the user out-of-the-box. JavaScript support is currently on par with plain text support on many levels. That situation can change. That situation must change.

romainl committed:
Handle feedback

rhysd commented on runtime/ftplugin/javascript.vim:
```diff
@@ -26,13 +26,70 @@ setlocal comments=sO:*\ -,mO:*\ \ ,exO:*/,s1:/*,mb:*,ex:*/,://
+    let s:bin_dir = fnamemodify(s:bin_dir, ':p')
+
+    if $PATH !~ s:bin_dir
+        let $PATH = s:bin_dir .. ':' .. $PATH
```
I think this does not work on Windows since path separator on Windows is `;`.

romainl replied:
You are right.

romainl replied:
Fixed. Do you know by any chance if there are other separators I should consider?

rhysd replied:
There is no new commit in this branch (you might forget pushing the fix).

> Do you know by any chance if there are other separators I should consider?

AFAIK considering `:` and `;` is sufficient. I know that Plan9 uses NUL (`"\0"`) for path separator, but I don't think Vim supports it.

romainl replied:
There.

romainl committed:
Made $PATH manipulation Windows-friendly

brammool commented:
Changing the 'path' option should be OK, it changes where Vim looks for included files.

Changing $PATH is questionable, as it tells the system where to find executable files.  It's a known security issue to set it to the current directory.  If there are actually executables in a directory $PATH should already include it, dynamically changing $PATH based on an edited file is wrong.

romainl commented:
> Changing $PATH is questionable, as it tells the system where to find executable files.

Only within Vim.

> It's a known security issue to set it to the current directory.

This is not what is proposed, here. Besides, if an attacker wants to use one of the locally installed binary he can *and will* do so by specifying its path.

> If there are actually executables in a directory $PATH should already include it, dynamically changing $PATH based on an edited file is wrong.

For better or for worse, JavaScript coders expect the following:

- project has zero system-wide dependencies,
- project-local dependencies end up in a `node_modules` directory at the root of the project,
- project-local dependencies that install binaries, generally development dependencies, do so in `node_modules/.bin`.

`node_modules/.bin` is where all the compilers/transpilers/linters/test runners/whatever used during development reside and that directory has nothing to do in the system's `$PATH` because a) that directory is not supposed to be accessed globally, b) that would needlessly pollute the global `$PATH` (I have more than 100 JS projects on this machine), and c) it would cause nightmarish versioning issues (I'm in project FOO and I need the tools for project FOO, not those of project BAR). What's in there is only relevant to the current project and as no business being accessed globally.

My proposition is to change `$PATH` *in Vim*, not globally, so that all those tools, that are sadly a necessary part of any JS dev's work day, can be used seamlessly with built-in features like `&formatprg`, `&makeprg`, etc. without too much setup.

That would be the exact equivalent of `$ PATH=./node_modules/.bin vim`, but automatic.

Again, setting up Vim for JavaScript is currently either an infinite and somewhat fun rabbit hole for those of us with that kind of kink or something so daunting (the defaults are utterly bad) that they turn to mutually incompatible third-party plugins.

We can do better.

k-takata commented on runtime/ftplugin/javascript.vim:
```diff
@@ -50,8 +50,10 @@ let s:bin_dir = finddir('node_modules/.bin', '.;')
 if !empty(s:bin_dir)
     let s:bin_dir = fnamemodify(s:bin_dir, ':p')
 
+    let path_separator = (has('win64') || has('win32') || has('win16')) ? ';' : ':'
```
```suggestion
    let path_separator = has('win32') ? ';' : ':'
```
When `has('win64')` is true, `has('win32')` is also true.
The win16 support was dropped a few years ago.

romainl replied:
Thank you. Done.

romainl committed:
Simplify system check for $PATH separator

dkearns commented:
I think some of this is starting to go around in circles which is probably my fault for not just saying "no". :)

> > As it stands, editing a JS file anywhere on the filesystem potentially modifies PATH and does so by favouring commands installed in any discovered node_modules/.bin directory.
> 
> Only if that directory is found in a parent directory of the current file. This is, for better or for worse, how it is supposed to work in JS land.

It doesn't work like that though, does it?  They, at least, offer the indirection of requiring the use of `npx` or `npm run` etc.  So when using these commands there is an implicit user commitment to search for binaries in these directories.  I don't think the same thing can be said of simply editing a JS file with Vim.

The required presence of `node_modules/.bin` is certainly not a mitigating factor with regard to any security concerns.

The proposed feature is not entirely dissimilar to having the next version of Bash come preinstalled with `npx --shell-auto-fallback`...if it were `--shell-auto-shadow`.

> > Anyway, if we were to include something like this I think it would have to be guarded by an opt-in user config variable.
> 
> I would prefer not to but that would be acceptable.

I don't see how it's possible to come up with a reasonable solution without requiring it be opt in.  Security issues aside, it just wreaks too much havoc for a default.  As you haven't thought to provide an opt out mechanism I'm assuming you think it's harmless and it is not.

It's going to break one of my common Node.js related configuration strategies rather badly and another completely.  I find it unlikely that I'm just coincidentally the only user for whom this isn't going to work.  You're also, obviously, going to break things for anyone employing a shim based strategy or using tools like Volta, and so on.

FWIW, I just installed VSCode for the first time to see what it was doing and discovered that the ESLint plugin does automagically discover the local installation but prompts you to approve its use with a modal dialog on, I think, a per project basis.

> > The value for PATH needs to be reset using the undo_ftplugin mechanism.
> 
> Agreed, and it is.

To clarify, I'm referring to the `PATH` environment variable.  The prefixed component(s) are never removed.

To finish the thought, though, undo_ftplugin alone isn't going to be enough.  I think it'll need to be managed by autocmds and it might be a bit tricky to get right as `PATH` should probably be the same for `:!` `:shell` and `:terminal`.

> An environment where they don't have to do that because their editor/IDE or their plugins are smart enough to do it for them.

As implemented it's not nearly smart enough and it breaks `:make` in a multitude of ways.  At a minimum it should be as functional as running Vim within an environment where the compiler command is already available.

Relying on this feature now a JS file has to be edited before the compiler can be run.  Why can't I run it when opening the project directory, or from package.json or any unrelated filetype when I haven't first edited a JS file?  I do all of these things at times.

Are you proposing adding this for every filetype potentially related to a Node project?  It should really be extracted to a plugin that can be used across all relevant filetypes.  There are, of course, other lang environments where it could also be used.

However, I think an ftplugin is entirely the wrong context to be configuring this and there's no reason it shouldn't work with `:filetype off`.  The better way is, probably, as a distributed plugin (if you can convince Bram) that is opt-in and operates entirely in the compiler context.  Each relevant compiler could optionally load the plugin and use it to find the binary and set `makeprg`.  Perhaps this could override `CompilerSet` as I think there's a few third-party plugins that rely on the standard `CompilerSet` line.

> Adding dozens of non-globally-significant paths to the globally significant `$PATH` makes no sense whatsoever.

You're reading me too literally (I hope) and I would expect no one is doing what you describe.  It, obviously, wouldn't even work.  I just meant, at worst, something has to be configured at project creation time to allow an appropriate environment to be constructed when needed.

> Having your tools handle `$PATH` dynamically for you within the context of editing makes a lot more sense.

Having your tools dynamically handle creating environments for you further up the process tree makes much more sense.  They can then even be used in multiple editing contexts if that's your sole focus.  There's lots of ways to do it depending on your workflow and lots of existing tools.

I don't understand why you think it's so easy to configure inside Vim and so difficult to configure externally.  If you can do it programmatically inside Vim you can do it programmatically outside Vim.

I notice that `direnv`, for example, provide this specific configuration but they also include a note suggesting it's an ill-advised approach.

This is, however, getting way out of scope.  I think the feature is completely out of scope for a ftplugin where we're not even supposed to be altering global options.

> `npx` is a cool tool for one-off tasks but the resolution adds ~2s to each command so it is unusable for lint-on-write scenarios, for example, or even the occasional `gq`.

That's odd.  They boast about it having no overhead for installed binaries and I see nothing significant.

```
projects/javascript/vim-ftplugin via ⬢ v14.8.0 
❯ time npx eslint --version
v7.7.0

real    0m0.608s
user    0m0.596s
sys     0m0.065s
projects/javascript/vim-ftplugin via ⬢ v14.8.0 
❯ time ./node_modules/.bin/eslint --version
v7.7.0

real    0m0.534s
user    0m0.534s
sys     0m0.054s

```
I was considering adding a `nodejs_use_npx` config option for use by the relevant compilers.

> > I think there's more value in consistency of behaviour across all ftplugins even if that might seem overly conservative in some instances.
> 
> I think there is more value in providing as much value as possible to the user out-of-the-box. JavaScript support is currently on par with plain text support on many levels. That situation can change. That situation must change.

This is entering philosophical territory but I'm certainly not arguing the default plugin shouldn't be improved.  I just don't see some of these additions as improvements but rather personal configurations that should remain in `~/.vim`.  I'm a user too and they provide less than no value.  I have to actively work around them in special ways.

So this doesn't get bogged down indefinitely, could you remove the `PATH` modifying section so we can easily merge the rest of your changes and then open another issue that addresses just the general automagical discovery of compiler binaries?

romainl committed:
Remove automatic$PATH and compiler setup

romainl commented:
> So this doesn't get bogged down indefinitely, could you remove the `PATH` modifying section so we can easily merge the rest of your changes and then open another issue that addresses just the general automagical discovery of compiler binaries?

OK, the `$PATH` setup part is gone, as well as the automatic `:compiler`.

brammool commented:
Romain Lafourcade wrote:

> > Changing $PATH is questionable, as it tells the system where to find
> > executable files.
> 
> Only within Vim.

I don't get it.  Vim itself doesn't use $PATH much.  It is used for
job_start(), is that what you are doing this for?

A complication is that $PATH is global for Vim, thus it's not just
changed for the current buffer.  What if another filetype would want
another value for $PATH?  At least it's unexpected for me that $PATH
changes just because I edited a .js file.

> > It's a known security issue to set it to the current directory.
> 
> This is not what is proposed, here.
> 
> > If there are actually executables in a directory $PATH should
> > already include it, dynamically changing $PATH based on an edited
> > file is wrong.
> 
> For better or for worse, JavaScript coders expect the following:
> 
> - project has zero system-wide dependencies,
> - project-local dependencies end up in a `node_modules` directory at
> the root of the project,
> - project-local dependencies that install binaries, generally
> development dependencies, do so in `node_modules/.bin`.

Where do these binaries come from?

> `node_modules/.bin` is where all the
> compilers/transpilers/linters/test runners/whatever used during
> development reside and that directory has nothing to do in the
> system's `$PATH` because a) that directory is not supposed to be
> accessed globally, b) that would needlessly pollute the global `$PATH`
> (I have more than 100 JS projects on this machine), and c) it would
> cause nightmarish versioning issues (I'm in project FOO and I need the
> tools for project FOO, not those of project BAR). What's in there is
> only relevant to the current project and as no business being accessed
> globally.

This sounds like it's not related to the filetype but to the compiler.

> My proposition is to change `$PATH` *in Vim*, not globally, so that
> all those tools, that are sadly a necessary part of any JS dev's work
> day, can be used seamlessly with built-in features like `&formatprg`,
> `&makeprg`, etc. without too much setup.
> 
> That would be the exact equivalent of `$ PATH=./node_modules/.bin
> vim`, but automatic.
> 
> Again, setting up Vim for JavaScript is currently either an infinite
> and somewhat fun rabbit hole for those us with that kind of kink or
> something so daunting (the defaults are utterly bad) that they turn to
> mutually incompatible third-party plugins.

I am worried that changing $PATH without the user's consent opens up a
security hole.  Security can be exchangable with convenience, like
asking for a password is annoying but requirement for safety.  It's not
an easy balance, I tend to stay on the safe side.

-- 
"Marriage is when a man and woman become as one; the trouble starts
when they try to decide which one"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

romainl commented:
> > Changing $PATH is questionable, as it tells the system where to find > executable files. Only within Vim.
> I don't get it. Vim itself doesn't use $PATH much. It is used for job_start(), is that what you are doing this for?

No, I do it to allow the user to use those binaries for linting/testing (via :compiler foo, :make %, and possibly autocommands) or for formatting (via &formatprg and gq, etc.). Basically, to make the built-in features easier to use than they currently are.

> A complication is that $PATH is global for Vim, thus it's not just changed for the current buffer. What if another filetype would want another value for $PATH?

The exact same thing that happens in your shell when you want to use /usr/bin/foo or /opt/local/bin/bar: both /usr/bin and /opt/local/bin are in your $PATH so you can simply use foo and bar without worrying about where they are.

$PATH is not replaced, one segment is prepended to it.

> At least it's unexpected for me that $PATH changes just because I edited a .js file.

You typically don't even think about `$PATH` when you edit a `.js` file as the value is totally irrelevant to what you are doing.

> Where do these binaries come from?

They are installed as part of $npm install or $ yarn

> This sounds like it's not related to the filetype but to the compiler.

Maybe, but there is no :formatter command similar to :compiler to set &formatprg.

> I am worried that changing $PATH without the user's consent opens up a security hole. Security can be exchangable with convenience, like asking for a password is annoying but requirement for safety. It's not an easy balance, I tend to stay on the safe side.

That is one of the reasons why people brush off built-in features so easily. They are neat on paper, but setting them up is too involved.

But the point is moot as I removed that part.

brammool commented:
I'll leave it to @dkearns to make further comments

dkearns commented:
> > So this doesn't get bogged down indefinitely, could you remove the `PATH` modifying section so we can easily merge the rest of your changes and then open another issue that addresses just the general automagical discovery of compiler binaries?
> 
> OK, the `$PATH` setup part is gone, as well as the automatic `:compiler`.

Thank you.  The only thing left to do is clean up the buffer local variables in `b:undo_ftplugin`.

Again, if you think the `includeexpr` you were working on is generally useful but imperfect we could add that too.

I was just reading about the changes to `npx` for [npm v7](https://blog.npmjs.org/post/626173315965468672/npm-v7-series-beta-release-and-semver-major).  It should provide the ability to use `--prefix` and this might be enough to provide a simple solution if a reasonable scheme can be developed to trigger the project prefix changes.

romainl committed:
Extend b:undo_ftplugin

romainl commented:
> Thank you. The only thing left to do is clean up the buffer local variables in `b:undo_ftplugin`.

Done.

romainl commented:
> Again, if you think the `includeexpr` you were working on is generally useful but imperfect we could add that too.

No, I'm not comfortable with doing so at the moment because of the design issues I mentioned earlier and their impact on performance. I would rather open a separate PR for that as it is quite the can of worms.

dkearns commented on runtime/ftplugin/javascript.vim:
```diff
@@ -65,7 +65,9 @@ let &l:define =
-let b:undo_ftplugin = "setl fo< ofu< com< cms< sua< su< def< pa<"
+let b:undo_ftplugin =
+            \ "setl fo< ofu< com< cms< sua< su< def< pa<"
+            \ .. "| unlet b:browsefilter b:match_ignorecase b:match_words"
```
Can I get you to add a `!` to `unlet`?  It otherwise causes errors if those variables weren't set.

Thanks.

romainl committed:
Banged :unlet

dkearns commented:
Thanks @romainl 

@brammool Can you please include this with your next runtime files update? Thanks.

brammool commented:
Thanks for the review.  I'll include it now.
